unblob/handlers/archive/qnap/qnap_nas.py

```diff
@@ -0,0 +1,203 @@
+    try:
+        header.device_id.decode("utf-8")
+        header.file_version.decode("utf-8")
+        header.firmware_date.decode("utf-8")
+        header.revision.decode("utf-8")
+    except UnicodeDecodeError:
+        return False
+    return True
+
+
+def _hyperscan_match(
+    context: QTSSearchContext, pattern_id: int, offset: int, end: int
+) -> Scan:
+    if offset < context.start_offset:
+        return Scan.Continue
+    struct_parser = StructParser(C_DEFINITIONS)
+    context.file.seek(offset, io.SEEK_SET)
+    header = struct_parser.parse("qnap_header_t", context.file, Endian.LITTLE)
+    logger.debug("qnap_header_t", header=header)
+
+    if is_valid_header(header):
+        context.end_block_offset = context.file.tell()
+        return Scan.Terminate
+    return Scan.Continue
+
+
+def build_stream_end_scan_db(pattern_list):
+    return StreamDatabase(
+        *(Pattern(p.as_regex(), Flag.SOM_LEFTMOST, Flag.DOTALL) for p in pattern_list)
+    )
+
+
+hyperscan_stream_end_magic_db = build_stream_end_scan_db(FOOTER_PATTERN)
+
+
+class QnapExtractor(Extractor):
+    def __init__(self):
+        self._struct_parser = StructParser(C_DEFINITIONS)
+
+    def extract(self, inpath: Path, outdir: Path):
+        outpath = outdir.joinpath(f"{inpath.name}.decrypted")
+        with File.from_path(inpath) as file:
+            file.seek(-FOOTER_LEN, io.SEEK_END)
+            header = self._struct_parser.parse("qnap_header_t", file, Endian.LITTLE)
+            end_offset = file.tell()
+            filesize = end_offset
+            cryptor = Cryptor(SECRET + header.file_version.decode("utf-8")[0])
+            with outpath.open("wb") as outfile:
+                for chunk in iterate_file(file, 0, header.encrypted_len, 1024):
+                    outfile.write(cryptor.decrypt_chunk(chunk))
+                for chunk in iterate_file(
+                    file,
+                    header.encrypted_len,
+                    filesize - FOOTER_LEN - header.encrypted_len,
+                    1024,
+                ):
+                    outfile.write(chunk)
+
+
+class QnapHandler(StructHandler):
+    NAME = "qnap_nas"
+
+    PATTERNS = [
+        HexString("F5 7B 47 03"),
+    ]
+    C_DEFINITIONS = C_DEFINITIONS
+    HEADER_STRUCT = "qnap_header_t"
+    EXTRACTOR = QnapExtractor()
+
+    def calculate_chunk(self, file: File, start_offset: int) -> Optional[ValidChunk]:
+        context = QTSSearchContext(
+            start_offset=start_offset, file=file, end_block_offset=-1
+        )
+
+        try:
+            scanner = hyperscan_stream_end_magic_db.build(context, _hyperscan_match)
+            stream_scan(scanner, file)
+        except Exception as e:
+            logger.debug(
+                "Error scanning for QNAP patterns",
+                error=e,
+            )
+
+        if context.end_block_offset > 0:
+            return ValidChunk(
+                start_offset=start_offset, end_offset=context.end_block_offset
+            )
```

Comment:
Please add some comment about the logic being applied here. Just explains what you're doing at a high level.